### ISSUE TITLE
TileLayer Inconsistent Z Calculation

### ISSUE THREAD

**Issue by mcost45** [visgl/deck.gl#5811]
#### Description
Tile-layer uses tileset-2d, which in turn uses the getTilesIndices function within tile-layer's util.js. GetTileIndices calculates the z value from the viewport zoom, which ends up getting passed into the getTileData({x, y, z}) prop for retreiving images:
`const TILE_SIZE = 512;`
`let z = Math.round(viewport.zoom + Math.log2(TILE_SIZE / tileSize));`

Because of this, I ran into a confusing and inconsistent problem with tiled layers behaving differently with tileSizes such as 256 (as in using two separate tiled image sources that use different tile image dimensions, e.g. one has images of 256x256, and the other has images of 512x512), being given the wrong z value to fetch data at, whereas other tiled layer sources would work properly.

For example, if the TileLayer view initialises with a zoom of -1.4 (0 being fully zoomed in & max-resolution), with a tileSize of 256, z will be calculated as `Math.round(1.4 +  Math.log2(512/256))` => 0. Then the TileLayer will fetch from z 0 at a zoom of -1.4, when it should instead be fetching from z -1 - displaying incorrect/stretched tiles on screen.

Conversely, with a tileSize of 512, the TileLayer would fetch the correct level tiles for the zoom as the z calculated from a -1.4 zoom would be -1, not 0.

My work around now involves applying an offset to the z value:
`const decimals = Math.floor(this.viewStates.main.zoom) - this.viewStates.main.zoom;`
`const offset = decimals  < -0.5 ? 1 : 0;`
`const DZIResolutionLevelAtZoomZero = DZIMaxLevel - offset;`

`getTileData(DZIResolutionLevelAtZoomZero + z, x, y)`

#### Expected Behavior
Unless I'm missing something, TileLayers should be able to consistently calculate and pass the correct z value regardless of tileSize, without needing to correct with an offset.

#### Repro Steps
Use TileLayer data sources with tileSizes of 256 and 512 to see the difference in how z get's calculated.

#### Environment
Framework Version: 8.4.16
Browser Version: Chrome 90.0.4430.212
OS: Windows 10.0.19042 Build 19042

#### Logs
None

**Comment by Pessimistress** [comment]
This is intentional. [tileSize](https://deck.gl/docs/api-reference/geo-layers/tile-layer#tilesize) defines the target pixel width and height of each tile when rendered, so smaller `tileSize` means the layer needs to fetch more tiles i.e. at a higher zoom. The whole purpose of this prop is to provide the "offset" that you mentioned.

There is no expectation that `tile.z` matches viewport zoom, especially in tilted maps. If setting `tileSize` to 512 yields the correct z in your use case, then that is the value you should use.

**Comment by mcost45** [comment]
> This is intentional. [tileSize](https://deck.gl/docs/api-reference/geo-layers/tile-layer#tilesize) defines the target pixel width and height of each tile when rendered, so smaller `tileSize` means the layer needs to fetch more tiles i.e. at a higher zoom. The whole purpose of this prop is to provide the "offset" that you mentioned.
> 
> There is no expectation that `tile.z` matches viewport zoom, especially in tilted maps. If setting `tileSize` to 512 yields the correct z in your use case, then that is the value you should use.

@Pessimistress Sorry, to better clarify I mean using two separate/different tiled image sources - one that has tiles of 256x256 images, and the other that has tiles of 512x512 images - and these are the pixel dimensions these tiles should be rendered at. I understand the point of the tileSize, and I'm not saying the issue is that any tileSize number should just work with any dimension tiles, but rather the issue is that there is an inconsistency in how the z value gets calculated between tiled image sources that do have different tile sizes - to me this issue still stands.

Without applying my own offset (not the same function as the tileSize offset), a 256x256 tiled image source will be one zoom level off where it should be, vs when using the 512x512 tiled image source, will fetch the correct zoom value (as per my example) - and that is why my own offset - in the z level being fetched -  becomes necessary, unrelated to the 'offset from tileSize'.

**Comment by Pessimistress** [comment]
Sounds like your issue is rather that `tile.z` does not match `tile.bbox` in your tile set. The TileLayer uses [OSM tile index](https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames) and z=0 is a single tile covering the entire Web Mercator world. If that is not how your tileset defines its z levels, then what you are doing is necessary - provide your own `getTileData` and remap the tile index before fetching.

**Comment by mcost45** [comment]
@Pessimistress In this case I am not using a geospatial view, but an orthographic view. [To me the docs imply that tilelayers within an orthographic view will not use the OSM tile index](https://deck.gl/docs/api-reference/geo-layers/tile-layer#indexing-systeml) - or is it still effectively the same? 

My tile sources are based on the deep zoom image format (DZI) - starting from full resolution, then repeatedly split into tiles of given tileSize (from top-left), then halve dimensions and split into tiles again, until 1px is reached. So far I have not ran into any problems with the 512 tileSize source, only needing to map z: 0 to the highest available resolution level, z: -1 to the second highest resolution level, etc. I will look into if it's a mapping issue further.

**Comment by Pessimistress** [comment]
See [this example](https://deck.gl/examples/tile-layer-non-geospatial/)

**Comment by mcost45** [comment]
@Pessimistress Yep, I used this example as a reference for my own code. I use the exact same method for offsetting z (e.g. `${15 + z}/${x}_${y}.jpeg`) to make it max at the largest resolution level, and also the sublayer bitmap clamping - which helped in getting the 512px tile set working easily. It was only when I changed to test a tileset with a 256px tilesize that I noticed it was always one zoom level off. Without further mapping, this example (and my own code) will only work with 512px.

My 256px tile set is also DZI like the example, and applies the same offset to the fetching z value. For DZI changing the tilesize also does not effect its generated levels - because they only depend on how many times its width needs to be divided by 2 before it reaches 1px. However I still need to apply an extra z offset - only using the offset in the example (adding the max resolution level index '15' to the fetched z value fetches tiles one level off what would be expected) - which is not needed in the case of my 512px (and moon example) source.

**Comment by Pessimistress** [comment]
Perhaps @ilan-gold can share more insights on how the DeepZoom system works

**Comment by ilan-gold** [comment]
@mcost45 You are correct in your analysis if I understand you correctly.  In order to calculate the offset correctly you can use something like: https://github.com/hms-dbmi/viv/blob/master/src/layers/MultiscaleImageLayer/MultiscaleImageLayer.js#L108-L112
There is a comment there about why it works and a [link to the line of code here](https://github.com/visgl/deck.gl/blob/26635d301171387c5c0509b5a099e151cf3de6f9/modules/geo-layers/src/tile-layer/utils.js#L183) that necessitates it.

@Pessimistress I can add some docs for this if you'd like, or add a fix - either way.  What @mcost45 is saying is similar to what I was saying in #4616 about the zoom-level being a little wonky.  The reason I that I made the change that necessitated #4616 was because previously the tile fetching mechanism was doing what @mcost45 described but in the geospatial context.  That is, it used to chronically underfetch tiles resulting in a blurry map.  However my fix resulted in fetching too aggressively.

So I see two options here:
1. Change https://github.com/visgl/deck.gl/blob/26635d301171387c5c0509b5a099e151cf3de6f9/modules/geo-layers/src/tile-layer/utils.js#L183 based on the type of viewport (geospatial or not), which would be a "breaking change" of sorts, although it's extremely possible I am the only consumer of this feature/fix/change besides now @mcost45.
2. Add documentation about this to `TileLayer` explaining how to get the correct zooming mechanism.
3. Something else?

Hope this helps!

**Comment by ilan-gold** [comment]
P.S I am happy (and volunteering) to do any and all changes here.

**Comment by mcost45** [comment]
Thanks to you both! Definitely helpful

**Comment by Pessimistress** [comment]
Thanks @ilan-gold ! I'm open to this adjustment in 8.5. For the long term, should we aim/advertise this layer to use DZI as the non-geo indexing standard? Is there anything more we need to do to claim full compatibility? What are the alternative standards?

**Comment by ilan-gold** [comment]
@Pessimistress If I understand you correctly, then yes, I believe that this indexing is fairly standard.  All of our TIFF file pyramids appear to have this same "issue."  I don't know about different standards to be honest - I have seen things like downsampling by numbers other than 2 but other than that I have not encountered anything else.  I can do some poking around, though - I'll try to figure out what the "indexing" really is so we can document it and have good motivation for the change.

**Comment by ilan-gold** [comment]
@Pessimistress Got around to looking at this, looks like we would need to revert a lot of https://github.com/visgl/deck.gl/pull/4616/files to do this.  I think it's worth it though.  I'll open a PR and we can reject if we don't like how it comes out.

**Comment by peaceshi** [comment]
I also has this issue and seems like the tiles layer has performance issue when zooming or drawing.
see smaple site
 https://yuanshen.site/v3/

**Comment by ilan-gold** [comment]
@Pessimistress I created [a branch](https://github.com/visgl/deck.gl/compare/master...ilan-gold:infovis_tile_offset?expand=1) with the necessary changes/reversions.

One thing I ran into while fixing this issue was that I actually had been making use of this ability to change the "zoom offset" arbitrarily for a nicer visual when the `modelMatrix` has a scaling factor.  Specifically these lines
https://github.com/hms-dbmi/viv/blob/e36160c4e9934d2eef7ed139fb204e2a18cac39a/src/layers/MultiscaleImageLayer/MultiscaleImageLayer.js#L186-L188
allows us to solve this issue: https://github.com/hms-dbmi/viv/issues/349

Maybe I could make a first PR from the above referenced branch and then another PR for some sort of offset prop?  I do think this feature is nice to have but I am not sure what the best way to use it is.  Maybe the solution to this whole issue of "TileLayer Inconsistent Z Calculation" is to just document it very well.  It is nice to be able to customize this sort of behavior.
